flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/RegionPartitionReleaseStrategyTest.java

```diff
@@ -0,0 +1,147 @@
+			PipelinedRegion.from(onlySourceVertexId),
+			PipelinedRegion.from(onlyIntermediateVertexId, onlySinkVertexId));
+
+		final RegionPartitionReleaseStrategy regionPartitionReleaseStrategy = new RegionPartitionReleaseStrategy(testingSchedulingTopology, pipelinedRegions);
+
+		regionPartitionReleaseStrategy.vertexFinished(onlyIntermediateVertexId);
+		final List<IntermediateResultPartitionID> partitionsToRelease = regionPartitionReleaseStrategy.vertexFinished(onlySinkVertexId);
+		assertThat(partitionsToRelease, contains(onlySourceResultPartitionId));
+	}
+
+	@Test
+	public void notReleasePartitionsIfDownstreamRegionIsNotFinished() {
+		final List<TestingSchedulingExecutionVertex> producers = testingSchedulingTopology.addExecutionVertices().finish();
+		final List<TestingSchedulingExecutionVertex> consumers = testingSchedulingTopology.addExecutionVertices().withParallelism(2).finish();
+		testingSchedulingTopology.connectAllToAll(producers, consumers).finish();
+
+		final ExecutionVertexID onlyProducerVertexId = producers.get(0).getId();
+		final ExecutionVertexID consumerVertex1 = consumers.get(0).getId();
+		final ExecutionVertexID consumerVertex2 = consumers.get(1).getId();
+
+		final Set<PipelinedRegion> pipelinedRegions = pipelinedRegionsSet(
+			PipelinedRegion.from(onlyProducerVertexId),
+			PipelinedRegion.from(consumerVertex1, consumerVertex2));
+
+		final RegionPartitionReleaseStrategy regionPartitionReleaseStrategy = new RegionPartitionReleaseStrategy(testingSchedulingTopology, pipelinedRegions);
+
+		final List<IntermediateResultPartitionID> partitionsToRelease = regionPartitionReleaseStrategy.vertexFinished(consumerVertex1);
+		assertThat(partitionsToRelease, is(empty()));
+	}
+
+	@Test
+	public void toggleVertexFinishedUnfinished() {
+		final List<TestingSchedulingExecutionVertex> producers = testingSchedulingTopology.addExecutionVertices().finish();
+		final List<TestingSchedulingExecutionVertex> consumers = testingSchedulingTopology.addExecutionVertices().finish();
+		final List<TestingSchedulingResultPartition> resultPartitions = testingSchedulingTopology.connectPointwise(producers, consumers)
+			.withResultPartitionType(ResultPartitionType.PIPELINED).finish();
+
+		final ExecutionVertexID onlyProducerVertexId = producers.get(0).getId();
+		final ExecutionVertexID onlyConsumerVertexId = consumers.get(0).getId();
+		final IntermediateResultPartitionID onlyResultPartitionId = resultPartitions.get(0).getId();
+
+		final Set<PipelinedRegion> pipelinedRegions = pipelinedRegionsSet(
+			PipelinedRegion.from(onlyProducerVertexId),
+			PipelinedRegion.from(onlyConsumerVertexId));
+		final RegionPartitionReleaseStrategy regionPartitionReleaseStrategy = new RegionPartitionReleaseStrategy(testingSchedulingTopology, pipelinedRegions);
+
+		regionPartitionReleaseStrategy.vertexFinished(onlyConsumerVertexId);
+		regionPartitionReleaseStrategy.vertexUnfinished(onlyConsumerVertexId);
```

Comment:
This test would pass even if vertexUnfinished is a no-op.

Also, shouldn't we at some point mark the producer vertex as finished? A scenario where a consumer is finished without the producer being finished sounds a bit like undefined behavior.

Reply:
Good catch.

> Also, shouldn't we at some point mark the producer vertex as finished? A scenario where a consumer is finished without the producer being finished sounds a bit like undefined behavior.

We can but:
-  calls to `ExecutionGraph#updateState(TaskExecutionState state)` can overtake each other. Therefore it is to be expected that a consumer is finished (in our bookkeeping) before the producer.
-  I don't think it adds to the readability of the test